File: python/copass-management/src/copass_management/adapters/mcp.py
"""Optional MCP adapter.

The Python ``mcp`` package is the only dependency this module imports
— guarded behind a runtime check so the rest of ``copass-management``
stays MCP-agnostic. Backend Phase 3 reuses ``register_management_tools``
without pulling MCP.
"""

from __future__ import annotations

import json
from typing import Any, List, TYPE_CHECKING

from copass_core import CopassClient

from copass_management.registrar import (
    RegistrarOptions,
    ToolRegistration,
    register_management_tools,
)

if TYPE_CHECKING:
    # Avoid a hard import — MCP is optional.
    from mcp.server.fastmcp import FastMCP


def register_to_mcp_server(
    server: "FastMCP",
    client: CopassClient,
    options: RegistrarOptions,
) -> List[ToolRegistration]:
    """Wire every read-tool registration onto an MCP server.

    Uses ``server.add_tool(...)`` from the ``mcp`` Python package, which
    must be installed separately. The MCP SDK is the only dependency
    this adapter takes — the underlying registrar stays
    transport-agnostic so backend Phase 3 can reuse it without the SDK.
    """

    def _register(registration: ToolRegistration) -> None:
        async def _handler(**kwargs: Any) -> str:
            result = await registration.handler(kwargs)
            if isinstance(result, str):
                return result
            return json.dumps(result, default=str)

        # Stash the spec on the function so introspection tools can find it.
        _handler.__name__ = registration.name
        _handler.__doc__ = registration.description

        server.add_tool(
            _handler,
            name=registration.name,
            description=registration.description,
        )

    return register_management_tools(_register, client, options)


__all__ = ["register_to_mcp_server"]

File: typescript/packages/management/src/adapters/mcp.ts
import type { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import type { CopassClient } from '@copass/core';

import {
  registerManagementTools,
  type RegistrarOptions,
  type ToolRegistration,
} from '../registrar.js';

/**
 * Wire every read-tool registration onto an MCP server. The MCP SDK is the
 * only dependency this adapter takes — the underlying registrar stays
 * transport-agnostic so backend Phase 3 can reuse it without pulling MCP.
 */
export function registerToMcpServer(
  server: McpServer,
  client: CopassClient,
  options: RegistrarOptions,
): ToolRegistration[] {
  return registerManagementTools(
    (registration) => {
      // The MCP SDK accepts a Zod schema for `inputSchema`/`outputSchema`.
      // We hand it the compiled Zod from our spec corpus so MCP-side
      // validation matches the conformance contract exactly.
      server.registerTool(
        registration.name,
        {
          description: registration.description,
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          inputSchema: registration.inputZod as any,
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          outputSchema: registration.outputZod as any,
        },
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        (async (rawInput: unknown) => {
          const result = await registration.handler(rawInput);
          const isObjectResult =
            typeof result === 'object' && result !== null;
          return {
            content: [
              {
                type: 'text',
                text:
                  typeof result === 'string'
                    ? result
                    : JSON.stringify(result, null, 2),
              },
            ],
            ...(isObjectResult
              ? { structuredContent: result as Record<string, unknown> }
              : {}),
          };
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
        }) as any,
      );
    },
    client,
    options,
  );
}
